### PR TITLE
Update Pipelines folder structure

### PR DESCRIPTION
**Pull request by erohmensing**
## Overview
This is an update to the folder/file structure of our `pipelines` project.

The goal is through this change to improve readability and context sharing for new and existing developers.

However, This is **NOT** an update to any logic or code. That is out of scope for now.

**The goal is to merge this fast so we dont have too much rebase hell**

## Changes

These are the broad high level changes and why

### 1. The colocation of pipelines and commands
Currently and in the future we believe the pipeline and the command concepts map 1:1.

Ideally we would be able to merge the abstractions (like in `aircmd`) but for now we believe a good first step is to hold them in the same location.

This allows
1. The folder structure to show what commands are available
2. Create a clear pairing between the pipeline code that is run and the cli arguments that feed them

### 2. A top level folder for CLI entry points
Even though commands are located beside their respective pipelines the entry point to the airbyte-ci should be as close to the root as possible and visible 

This allows
1. any new developer know where to start tracing code.
2. A single area to collect CLI helpers like telemetry and custom click classes

### 3. The reduction of large files into smaller single purpose files
This is to improve reading and tracing code. 

We've split
1. `bases.py` into models
2. utils.py into helpers
3. environments.py into dagger/actions and dagger/containers

### 4. Steps get more generic as you go up the file structure
The idea is that there is a taxonomy of steps that should be separated based on their wish to be reused.

1. Step Base Class now found at `models/steps.py`
1. Generic Steps intended to be reused (e.g. `PoetryRunStep`, `SimpleDockerStep`) now found at `pipeline/steps`
2. Single Purpose General Steps (e.g. `GitPushChanges`) now found at `pipeline/steps`
3. Specialize Pipeline specific steps (e.g. `BuildConnectorDistributionTar`) found at `pipeline/connectors/build/steps`

Separating them on this taxonomy lets developers infer how wide spread their usage or if the step is intended to work in any pipeline vs just a specific set of pipelines

### 5. Containers separated from actions
Environments.py had a mix of two different type signatures
* **Container -> Container**: These are functions that install something, analogous to dockerfile commands 
* **Context|str|any -> Container**: These are functions that create/initialize a container, analogous to dockerfile base image

We broke this into `actions` and `containers` respectively. This is to make this distinction clearer at a glance and potentially lean on the fact that the developer may be used to a distinction like this if theyve used docker before

## File Structure
```bash
.
├── README.md
├── pipelines
│   ├── __init__.py
│   ├── cli
│   │   ├── airbyte_ci.py
│   │   ├── dagger_pipeline_command.py
│   │   ├── dagger_run.py
│   │   └── telemetry.py
│   ├── consts.py
│   ├── dagger
│   │   ├── __init__.py
│   │   ├── actions
│   │   │   ├── __init__.py
│   │   │   ├── connector
│   │   │   │   ├── hooks.py
│   │   │   │   └── normalization.py
│   │   │   ├── python
│   │   │   │   ├── __init__.py
│   │   │   │   ├── common.py
│   │   │   │   ├── pipx.py
│   │   │   │   └── poetry.py
│   │   │   ├── remote_storage.py
│   │   │   ├── secrets.py
│   │   │   └── system
│   │   │       ├── __init__.py
│   │   │       ├── common.py
│   │   │       └── docker.py
│   │   └── containers
│   │       ├── __init__.py
│   │       ├── internal_tools.py
│   │       ├── java.py
│   │       └── python.py
│   ├── hacks.py
│   ├── helpers
│   │   ├── __init__.py
│   │   ├── connectors
│   │   │   ├── __init__.py
│   │   │   ├── metadata_change_helpers.py
│   │   │   └── modifed.py
│   │   ├── gcs.py
│   │   ├── git.py
│   │   ├── github.py
│   │   ├── sentry_utils.py
│   │   ├── slack.py
│   │   ├── steps.py
│   │   └── utils.py
│   ├── internal_tools
│   │   └── internal.py
│   ├── models
│   │   ├── bases.py
│   │   ├── contexts.py
│   │   ├── reports.py
│   │   └── steps.py
│   └── pipeline
│       ├── __init__.py
│       ├── connectors
│       │   ├── __init__.py
│       │   ├── commands.py
│       │   ├── context.py
│       │   ├── pipeline.py
│       │   ├── builds
│       │   │   ├── __init__.py
│       │   │   ├── commands.py
│       │   │   ├── pipeline.py
│       │   │   └── steps
│       │   ├── bump_version
│       │   │   ├── __init__.py
│       │   │   ├── commands.py
│       │   │   └── pipeline.py
│       │   ├── format
│       │   │   ├── __init__.py
│       │   │   ├── commands.py
│       │   │   ├── pipeline.py
│       │   │   └── steps
│       │   ├── list
│       │   │   ├── __init__.py
│       │   │   ├── commands.py
│       │   │   └── pipeline.py
│       │   ├── migrate_to_base_image
│       │   │   ├── __init__.py
│       │   │   ├── commands.py
│       │   │   └── pipeline.py
│       │   ├── publish
│       │   │   ├── __init__.py
│       │   │   ├── commands.py
│       │   │   ├── context.py
│       │   │   └── pipeline.py
│       │   ├── reports.py
│       │   ├── test
│       │   │   ├── __init__.py
│       │   │   ├── commands.py
│       │   │   ├── pipeline.py
│       │   │   └── steps
│       │   └── upgrade_base_image
│       │       ├── __init__.py
│       │       ├── commands.py
│       │       └── pipeline.py
│       ├── metadata
│       │   ├── __init__.py
│       │   ├── commands.py
│       │   └── pipeline.py
│       ├── steps
│       │   ├── __init__.py
│       │   ├── docker.py
│       │   ├── git.py
│       │   ├── gradle.py
│       │   ├── no_op.py
│       │   └── poetry.py
│       └── test
│           ├── __init__.py
│           ├── commands.py
│           └── pipeline.py
├── poetry.lock
├── pyproject.toml
└── tests
    ├── __init__.py
    ├── conftest.py
    ├── test_actions
    │   └── test_environments.py
    ├── test_bases.py
    ├── test_builds
    │   ├── dummy_build_customization.py
    │   └── test_python_connectors.py
    ├── test_commands
    │   ├── __init__.py
    │   └── test_groups
    │       ├── __init__.py
    │       └── test_connectors.py
    ├── test_gradle.py
    ├── test_publish.py
    ├── test_steps
    │   ├── __init__.py
    │   └── test_simple_docker_step.py
    ├── test_tests
    │   ├── __init__.py
    │   ├── test_common.py
    │   └── test_python_connectors.py
    ├── test_utils.py
    └── utils.py

36 directories, 105 files

```

### CODE REVIEW

**Comment by bnchrch** [comment]
File structure
```bash
.
├── README.md
├── pipeline_reports
│   ├── airbyte-ci
│   │   ├── connectors
│   │   │   ├── build
│   │   │   │   └── manual
│   │   │   ├── format
│   │   │   │   └── manual
│   │   │   ├── publish
│   │   │   │   └── manual
│   │   │   └── test
│   │   │       └── manual
│   │   └── manual
│   │       └── test_branch
│   │           └── 0
│   └── airbyte-ci-internal
│       └── connectors
│           └── publish
│               └── manual
├── pipelines
│   ├── __init__.py
│   ├── cli
│   │   ├── airbyte_ci.py
│   │   ├── dagger_pipeline_command.py
│   │   ├── dagger_run.py
│   │   └── telemetry.py
│   ├── consts.py
│   ├── dagger
│   │   ├── __init__.py
│   │   ├── actions
│   │   │   ├── __init__.py
│   │   │   ├── connector
│   │   │   │   ├── hooks.py
│   │   │   │   └── normalization.py
│   │   │   ├── internal_tools.py
│   │   │   ├── python
│   │   │   │   ├── __init__.py
│   │   │   │   ├── common.py
│   │   │   │   ├── pipx.py
│   │   │   │   └── poetry.py
│   │   │   ├── remote_storage.py
│   │   │   ├── secrets.py
│   │   │   └── system
│   │   │       ├── __init__.py
│   │   │       ├── common.py
│   │   │       └── docker.py
│   │   └── containers
│   │       ├── __init__.py
│   │       ├── java.py
│   │       └── python.py
│   ├── hacks.py
│   ├── helpers
│   │   ├── __init__.py
│   │   ├── connectors
│   │   │   ├── __init__.py
│   │   │   ├── metadata_change_helpers.py
│   │   │   └── modifed.py
│   │   ├── gcs.py
│   │   ├── git.py
│   │   ├── github.py
│   │   ├── sentry_utils.py
│   │   ├── slack.py
│   │   ├── steps.py
│   │   └── utils.py
│   ├── internal_tools
│   │   └── internal.py
│   ├── models
│   │   ├── bases.py
│   │   ├── contexts.py
│   │   ├── reports.py
│   │   └── steps.py
│   └── pipeline
│       ├── __init__.py
│       ├── connectors
│       │   ├── __init__.py
│       │   ├── builds
│       │   │   ├── __init__.py
│       │   │   ├── commands.py
│       │   │   ├── pipeline.py
│       │   │   └── steps
│       │   ├── bump_version
│       │   │   ├── __init__.py
│       │   │   ├── commands.py
│       │   │   └── pipeline.py
│       │   ├── commands.py
│       │   ├── context.py
│       │   ├── format
│       │   │   ├── __init__.py
│       │   │   ├── commands.py
│       │   │   ├── pipeline.py
│       │   │   └── steps
│       │   ├── list
│       │   │   ├── __init__.py
│       │   │   ├── commands.py
│       │   │   └── pipeline.py
│       │   ├── migrate_to_base_image
│       │   │   ├── __init__.py
│       │   │   ├── commands.py
│       │   │   └── pipeline.py
│       │   ├── pipeline.py
│       │   ├── publish
│       │   │   ├── __init__.py
│       │   │   ├── commands.py
│       │   │   ├── context.py
│       │   │   └── pipeline.py
│       │   ├── test
│       │   │   ├── __init__.py
│       │   │   ├── commands.py
│       │   │   ├── pipeline.py
│       │   │   └── steps
│       │   └── upgrade_base_image
│       │       ├── __init__.py
│       │       ├── commands.py
│       │       └── pipeline.py
│       ├── metadata
│       │   ├── __init__.py
│       │   ├── commands.py
│       │   └── pipeline.py
│       ├── steps
│       │   ├── __init__.py
│       │   └── git.py
│       └── test
│           ├── __init__.py
│           ├── commands.py
│           └── pipeline.py
├── poetry.lock
├── pyproject.toml
├── pytest.ini
└── tests
    ├── __init__.py
    ├── conftest.py
    ├── test_actions
    │   └── test_environments.py
    ├── test_bases.py
    ├── test_builds
    │   ├── dummy_build_customization.py
    │   └── test_python_connectors.py
    ├── test_commands
    │   ├── __init__.py
    │   └── test_groups
    │       ├── __init__.py
    │       └── test_connectors.py
    ├── test_gradle.py
    ├── test_publish.py
    ├── test_steps
    │   ├── __init__.py
    │   └── test_simple_docker_step.py
    ├── test_tests
    │   ├── __init__.py
    │   ├── test_common.py
    │   └── test_python_connectors.py
    ├── test_utils.py
    └── utils.py

54 directories, 100 files

```

File: airbyte-ci/connectors/pipelines/pipelines/dagger/actions/secrets.py [hunk 1]
```diff
@@ -42,9 +41,12 @@ async def download(context: ConnectorContext, gcp_gsm_env_variable_name: str = "
     Returns:
         Directory: A directory with the downloaded secrets.
     """
+    # temp - fix circular import
```
Comment (erohmensing) [inline airbyte-ci/connectors/pipelines/pipelines/dagger/actions/secrets.py:44]: @bnchrch this is the one circular import I just couldn't figure out how to implement in a decent way. It's also imported [here](https://github.com/airbytehq/airbyte/pull/31525/files#diff-b84200253d5dee854ad304834122ab7eb89978982d6d081a58337b9cf1dd03efR83). Maybe you can take a crack at it

**Review by alafanechere — APPROVED** [review]
Very cool changes! Thank you!!!!!!
Just minor suggestions.
Feel free to merge if the unit test pass.
A pre-release + a java and python connector test run would also be reassuring (feel free to ignore if you tested this locally).

File: airbyte-ci/connectors/pipelines/pipelines/pipeline/connectors/list/commands.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+from connector_ops.utils import console
+from pipelines.cli.dagger_pipeline_command import DaggerPipelineCommand
+from pipelines.pipeline.connectors.commands import connectors
+from rich.table import Table
+from rich.text import Text
+
+
+@connectors.command(cls=DaggerPipelineCommand, help="List all selected connectors.")
+@click.pass_context
+def list(
```
[the remark this answers is not in the document]
Comment (erohmensing) [reply airbyte-ci/connectors/pipelines/pipelines/pipeline/connectors/list/commands.py:15]: good point - for what its worth, the test command (running tests for airbyte ci) is not a pipeline either, nor will the new format be. I don't think these are utils though, maybe it is just a naming problem. 

@bnchrch you talked about pipelines, not commands, being the first class citizens here, somehow, right? I can't remember how that conversation went

File: airbyte-ci/connectors/pipelines/pipelines/dagger/containers/__init__.py
[the remark this answers is not in the document]
Comment (erohmensing) [reply airbyte-ci/connectors/pipelines/pipelines/dagger/containers/__init__.py:1]: Yes this is what we landed on! Do you think it makes sense?

File: airbyte-ci/connectors/pipelines/pipelines/internal_tools/internal.py
[the remark this answers is not in the document]
Comment (erohmensing) [reply airbyte-ci/connectors/pipelines/pipelines/internal_tools/internal.py:1]: Definitely- done!

File: airbyte-ci/connectors/pipelines/pipelines/dagger/containers/__init__.py
[the remark this answers is not in the document]
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/dagger/containers/__init__.py:1]: Definitely

[the remark this answers is not in the document]
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/dagger/containers/__init__.py:1]: Correct!

**Comment by erohmensing** [comment]
Current dependencies on/for this PR:
* master
  * **PR #31525** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/31525" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/airbytehq/airbyte/31525?utm_source=stack-comment).

File: airbyte-ci/connectors/pipelines/pipelines/internal_tools/internal.py
[the remark this answers is not in the document]
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/internal_tools/internal.py:1]: Yup!

File: airbyte-ci/connectors/pipelines/pipelines/pipeline/connectors/list/commands.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (erohmensing) [reply airbyte-ci/connectors/pipelines/pipelines/pipeline/connectors/list/commands.py:15]: > Currently Im thinking we leave it as is but perhaps find a new name for pipeline
> maybe even airbyte-ci?

This was going to be my exact suggestion, i say we go for it 😄

**Comment by erohmensing** [comment]
Some successful javapython connector test runs here: https://github.com/airbytehq/airbyte/pull/31586/ (after fixing something so definitely worth testing!) 

Successful prerelease here: https://github.com/airbytehq/airbyte/actions/runs/6567115837

File: airbyte-ci/connectors/pipelines/pipelines/pipeline/connectors/builds/__init__.py
Comment (pedroslopez) [inline airbyte-ci/connectors/pipelines/pipelines/pipeline/connectors/builds/__init__.py:1]: nit: why is this folder plural? `builds`? I think I would've expected it to line up with the command `build` if i understand the explanation correctly

**Review by pedroslopez — APPROVED** [review]
Nice! I kept having to hop around and this feels like would make things much clearer!

File: airbyte-ci/connectors/pipelines/pipelines/pipeline/connectors/builds/__init__.py
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/pipeline/connectors/builds/__init__.py:1]: Good catch! created an exception in our gitignore and renamed

Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/pipeline/connectors/builds/__init__.py:1]: just kidding, that still messes with our pipeline exclude logic. 

went for `build_image` instead

**Pull request merged**